File: src/workspace/src/logger/src/loggerText.cpp
#include "loggerText.h"

LoggerText::LoggerText(std::string & logFolder, std::string separator):LoggerBase(logFolder), separator(separator){
	readVitalsMsgFile();
}

LoggerText::~LoggerText(){
	finalize();
}

/*
 * Creates and opens the logger files with a proper timestamp and headers into a temporary location
 */
void LoggerText::init(){
	//Make sure environment is sane and that logging is enabled
	if(bootstrappedGnssTime && loggerEnabled){
		fileRotationMutex.lock();

		//Make sure the files are not already opened...
		if(!gnssOutputFile && !imuOutputFile && !sonarOutputFile && !lidarOutputFile && !rawGnssOutputFile.is_open() && !rawGnssOutputFile.is_open() && !vitalsOutputFile && !speedOutputFile){

			std::string dateString = TimeUtils::getStringDate();
			
			initGnssFile(dateString);
			initImuFile(dateString);
			initSonarFile(dateString);
			initLidarFile(dateString);
			initSpeedFile(dateString);
			initVitalsFile(dateString);
			
			rawGnssFileName = dateString + this->fileExtensionForGpsDatagram;
			rawGnssOutputFile.open(tmpLoggingFolder + "/" + rawGnssFileName,std::ios::binary|std::ios::trunc);
			
			if( !rawGnssOutputFile.good()){
				throw std::invalid_argument("Couldn't open raw gnss log file");
			}
			
			rawSonarFileName = dateString + this->fileExtensionForSonarDatagram;
			rawSonarOutputFile.open(tmpLoggingFolder + "/" + rawSonarFileName,std::ios::binary|std::ios::trunc);
			
			if( !rawSonarOutputFile.good()){
				throw std::invalid_argument("Couldn't open raw sonar log file");
			}
			
		}

		lastRotationTime = ros::Time::now();

		fileRotationMutex.unlock();
	}
}

void LoggerText::initGnssFile(std::string &dateString){
	//Open GNSS file
	gnssFileName = dateString + "_gnss.txt";

	std::string gnssFilePath = tmpLoggingFolder + "/"  + gnssFileName;

	gnssOutputFile= fopen(gnssFilePath.c_str(),"a");

	if(!gnssOutputFile){
		fileRotationMutex.unlock();
		throw std::invalid_argument(std::string("Couldn't open GNSS log file ") + gnssFileName);
	}
	
	fprintf(gnssOutputFile,"Timestamp%sLongitude%sLatitude%sEllipsoidalHeight%sStatus%sService\n",
									separator.c_str(),separator.c_str(),separator.c_str(),separator.c_str(),separator.c_str());
}

void LoggerText::initImuFile(std::string &dateString){
//Open IMU file
	imuFileName = dateString + "_imu.txt";

	std::string imuFilePath = tmpLoggingFolder + "/" + imuFileName;

	imuOutputFile = fopen(imuFilePath.c_str(),"a");

	if(!imuOutputFile){
		fileRotationMutex.unlock();
		throw std::invalid_argument(std::string("Couldn't open IMU log file ") + imuFileName);
	}
	
	fprintf(imuOutputFile,"Timestamp%sHeading%sPitch%sRoll\n",separator.c_str(),separator.c_str(),separator.c_str());
	
}

void LoggerText::initSonarFile(std::string &dateString){
	//Open sonar file
	sonarFileName = dateString + "_sonar.txt";

	std::string sonarFilePath = tmpLoggingFolder + "/" + sonarFileName;

	sonarOutputFile = fopen(sonarFilePath.c_str(),"a");

	if(!sonarOutputFile){
		fileRotationMutex.unlock();
		throw std::invalid_argument(std::string("Couldn't open sonar log file ") + sonarFileName);
	}
	
	fprintf(sonarOutputFile,"Timestamp%sDepth\n",separator.c_str());
}

void LoggerText::initLidarFile(std::string &dateString){
	//Open lidar file
	lidarFileName = dateString + "_lidar.txt";

	std::string lidarFilePath = tmpLoggingFolder + "/" + lidarFileName;

	lidarOutputFile = fopen(lidarFilePath.c_str(),"a");

	if(!lidarOutputFile){
		fileRotationMutex.unlock();
		throw std::invalid_argument(std::string("Couldn't open lidar log file ") + lidarFileName);
	}
	
	fprintf(lidarOutputFile,"Timestamp%sPoints\n",separator.c_str());
}

void LoggerText::initSpeedFile(std::string &dateString){
	//Open speed file
	speedFilename = dateString + "_speed.txt";

	std::string speedFilePath = tmpLoggingFolder + "/" + speedFilename;

	speedOutputFile = fopen(speedFilePath.c_str(),"a");

	if(!speedOutputFile){
		fileRotationMutex.unlock();
		throw std::invalid_argument(std::string("Couldn't open speed log file ") + speedFilename);
	}
	
	fprintf(speedOutputFile,"Timestamp%sSpeed\n",separator.c_str());
}

void LoggerText::initVitalsFile(std::string &dateString){
	//Open vitals file
	vitalsFilename = dateString + "_vitals.txt";

	std::string vitalsFilePath = tmpLoggingFolder + "/" + vitalsFilename;

	vitalsOutputFile = fopen(vitalsFilePath.c_str(),"a");


	if(!vitalsOutputFile){
		fileRotationMutex.unlock();
		throw std::invalid_argument(std::string("Couldn't open vital log file ") + vitalsFilename);
	}
	
	fprintf(vitalsOutputFile,"Timestamp%sCPU Temp%sCPU Used%sFREE RAM%sFREE HDD%sUptime%sHumidity%sSystem Temp%sSupply Voltage\n",
									separator.c_str(),separator.c_str(),separator.c_str(),separator.c_str(),separator.c_str(),separator.c_str(),separator.c_str(),separator.c_str());

}


/*
 * Closes the logging files and moves them to the web server's record directory to be accessible to download
 */

void LoggerText::finalize(){
	fileRotationMutex.lock();

	std::string newGnssPath;
	std::string newImuPath;
	std::string newSonarPath;
	std::string newLidarPath;
	std::string newRawGnssPath;
	std::string newRawSonarPath;
	std::string newSpeedPath;
	std::string newVitalsPath;
	
	if(gnssOutputFile){
		//close
		fclose(gnssOutputFile);
		gnssOutputFile = NULL;

		//move
		std::string oldPath = tmpLoggingFolder + "/"  + gnssFileName;
		newGnssPath = outputFolder + "/" + gnssFileName;
		rename(oldPath.c_str(), newGnssPath.c_str());
	}

	if(imuOutputFile){
		//close
		fclose(imuOutputFile);
		imuOutputFile = NULL;

		//move
		std::string oldPath = tmpLoggingFolder + "/"  + imuFileName;
		newImuPath = outputFolder + "/" + imuFileName;
		rename(oldPath.c_str(), newImuPath.c_str());
	}

	if(sonarOutputFile){
		//close
		fclose(sonarOutputFile);
		sonarOutputFile = NULL;

		//move
		std::string oldPath = tmpLoggingFolder + "/"  + sonarFileName;
		newSonarPath = outputFolder + "/" + sonarFileName;
		rename(oldPath.c_str(), newSonarPath.c_str());
	}
	
	if(lidarOutputFile){
		//close
		fclose(lidarOutputFile);
		lidarOutputFile = NULL;

		//move
		std::string oldPath = tmpLoggingFolder + "/"  + lidarFileName;
		newLidarPath = outputFolder + "/" + lidarFileName;
		rename(oldPath.c_str(), newLidarPath.c_str());
	}
	
	if(rawGnssOutputFile.is_open()){
		//close
		rawGnssOutputFile.close();
		
		std::string oldPath = tmpLoggingFolder + "/"  + rawGnssFileName;
		newRawGnssPath = outputFolder + "/" + rawGnssFileName;
		rename(oldPath.c_str(), newRawGnssPath.c_str());
	}
	
	if(rawSonarOutputFile.is_open()){
		//close
		rawSonarOutputFile.close();
		
		std::string oldPath = tmpLoggingFolder + "/"  + rawSonarFileName;
		newRawSonarPath = outputFolder + "/" + rawSonarFileName;
		rename(oldPath.c_str(), newRawSonarPath.c_str());
	}
	
	if(speedOutputFile){
		//close
		fclose(speedOutputFile);
		speedOutputFile = NULL;

		
		std::string oldPath = tmpLoggingFolder + "/"  + speedFilename;
		newSpeedPath = outputFolder + "/" + speedFilename;
		rename(oldPath.c_str(), newSpeedPath.c_str());
	}
	
	if(vitalsOutputFile){
		//close
		fclose(vitalsOutputFile);
		vitalsOutputFile = NULL;
		
		std::string oldPath = tmpLoggingFolder + "/"  + vitalsFilename;
		newVitalsPath = outputFolder + "/" + vitalsFilename;
		rename(oldPath.c_str(), newVitalsPath.c_str());
	}
	
	fileRotationMutex.unlock();

	std::string zipFilename = gnssFileName.substr(0, 17) + std::string(".zip");
	
	if(activatedTransfer){
	
		std::vector<std::string> files{newGnssPath, newImuPath, newSonarPath, newLidarPath, newRawGnssPath, newRawSonarPath, newSpeedPath, newVitalsPath};
	
		bool noError = compress(zipFilename, files);
		if(noError && can_reach_server()){
			transfer();
		}
	}
}

/* Rotates logs based on time */
void LoggerText::rotate(){
	if(bootstrappedGnssTime && loggerEnabled){

		ros::Time currentTime = ros::Time::now();

		if(currentTime.toSec() - lastRotationTime.toSec() > logRotationIntervalSeconds){
			ROS_INFO("Rotating logs");
			//close (if need be), then reopen files.
			finalize();
			init();
		}
	}
}


void LoggerText::gnssCallback(const sensor_msgs::NavSatFix& gnss){
	
	processGnssFix(gnss);
	updateGeofence(gnss);
	if(!bootstrappedGnssTime && gnss.status.status >= 0){
		bootstrappedGnssTime = true;
	}

	if(bootstrappedGnssTime && loggerEnabled){

		if(!gnssOutputFile){
			init();
		}
		else{
			//We will clock the log rotation check on the GNSS data since it's usually the slowest at 1Hz, which is plenty
			//TODO: we could throttle this if CPU usage becomes an issue
			rotate();
		}

		uint64_t timestamp = TimeUtils::buildTimeStamp(gnss.header.stamp.sec, gnss.header.stamp.nsec);

		if(timestamp > lastGnssTimestamp){
			fprintf(gnssOutputFile,"%s%s%.10f%s%.10f%s%.3f%s%d%s%d\n",
				TimeUtils::getTimestampString(gnss.header.stamp.sec, gnss.header.stamp.nsec).c_str(),
				separator.c_str(),
				gnss.longitude,
				separator.c_str(),
				gnss.latitude,
				separator.c_str(),
				gnss.altitude,
				separator.c_str(),
				gnss.status.status,
				separator.c_str(),
				gnss.status.service
			);

			lastGnssTimestamp = timestamp;
		}
	}
}

void LoggerText::imuCallback(const sensor_msgs::Imu& imu){
	if(bootstrappedGnssTime && loggerEnabled){
		double heading = 0;
		double pitch   = 0;
		double roll	= 0;

		uint64_t timestamp = TimeUtils::buildTimeStamp(imu.header.stamp.sec, imu.header.stamp.nsec);

		if(timestamp > lastImuTimestamp){
			
			imuTransform(imu, roll, pitch, heading);
			
			fprintf(imuOutputFile,"%s%s%.3f%s%.3f%s%.3f\n",TimeUtils::getTimestampString(imu.header.stamp.sec, imu.header.stamp.nsec).c_str(),separator.c_str(),heading,separator.c_str(),pitch,separator.c_str(),roll);

			lastImuTimestamp = timestamp;
		}
	}
}

void LoggerText::sonarCallback(const geometry_msgs::PointStamped& sonar){
	if(bootstrappedGnssTime && loggerEnabled){
		uint64_t timestamp = TimeUtils::buildTimeStamp(sonar.header.stamp.sec, sonar.header.stamp.nsec);

		if(timestamp > lastSonarTimestamp){
			/*fprintf(sonarOutputFile,"%s%s%.3f\n",TimeUtils::getTimestampString(sonar.header.stamp.sec, sonar.header.stamp.nsec).c_str(),separator.c_str(),sonar.point.z);*/
			if (this->enableGeofence && not this->insideGeofence){
				ROS_INFO("Not logging as we are outside active geofence");
			}
			else {
				fprintf(sonarOutputFile,"%s%s%.3f\n",TimeUtils::getTimestampString(sonar.header.stamp.sec, sonar.header.stamp.nsec).c_str(),separator.c_str(),sonar.point.z);
			}
			lastSonarTimestamp = timestamp;
		}
	}
}

void LoggerText::lidarCallBack(const sensor_msgs::PointCloud2& lidar){
	//ROS_INFO_STREAM("lidar callback : " << lidar.data.size() << "\n" );
	
	sensor_msgs::PointCloud lidarXYZ;
	sensor_msgs::convertPointCloud2ToPointCloud(lidar, lidarXYZ);
	
	std::vector<geometry_msgs::Point32> points = lidarXYZ.points;
	
	if(bootstrappedGnssTime && loggerEnabled){
		uint64_t timestamp = TimeUtils::buildTimeStamp(lidar.header.stamp.sec, lidar.header.stamp.nsec);
		
		if(timestamp > lastLidarTimestamp){
			fprintf(lidarOutputFile,"%s%s",TimeUtils::getTimestampString(lidar.header.stamp.sec, lidar.header.stamp.nsec).c_str(), separator.c_str());
			lastLidarTimestamp = timestamp;
			for(auto const& point : points){
				fprintf(lidarOutputFile,"%f %f %f%s", point.x, point.y, point.z, separator.c_str());
			}
			fprintf(lidarOutputFile,"\n");
		}
	}
}

void LoggerText::saveSpeed(const nav_msgs::Odometry& speed){
	if(bootstrappedGnssTime && loggerEnabled){
		uint64_t timestamp = TimeUtils::buildTimeStamp(speed.header.stamp.sec, speed.header.stamp.nsec);

		if(timestamp > lastSonarTimestamp){
			fprintf(speedOutputFile,"%s%s%.3f\n",TimeUtils::getTimestampString(speed.header.stamp.sec, speed.header.stamp.nsec).c_str(),separator.c_str(),speed.twist.twist.linear.y);
			lastSonarTimestamp = timestamp;
		}
	}
}

void LoggerText::saveVitals(const raspberrypi_vitals_msg::sysinfo& msg){
	
	if(bootstrappedGnssTime && loggerEnabled){
		uint64_t timestamp = TimeUtils::buildTimeStamp(msg.header.stamp.sec, msg.header.stamp.nsec);

		if( (timestamp - lastVitalsTimestamp) > (60 * 1000000) ){
			
			fprintf(vitalsOutputFile,"%s%s%.3f%s%.3f%s%.3f%s%.3f%s%.3f%s%.3f%s%.3f%s%.3f\n",
				TimeUtils::getTimestampString(msg.header.stamp.sec, msg.header.stamp.nsec).c_str(),
				separator.c_str(),
				msg.cputemp,
				separator.c_str(),
				msg.cpuload,
				separator.c_str(),
				msg.freeram,
				separator.c_str(),
				msg.freehdd,
				separator.c_str(),
				msg.uptime,
				separator.c_str(),
				msg.humidity,
				separator.c_str(),
				msg.temperature,
				separator.c_str(),
				msg.voltage
				);
			
			lastVitalsTimestamp = timestamp;
		}
	}
	
}

void LoggerText::readVitalsMsgFile(){

	std::ifstream file("/opt/Poseidon/src/workspace/src/raspberrypi_vitals_msg/msg/sysinfo.msg");
	if (!file.is_open()) {
		ROS_ERROR("Error opening file: /opt/Poseidon/src/workspace/src/raspberrypi_vitals_msg/msg/sysinfo.msg");
		return;
	}

	std::string line, firstCol, word;
	
	while (std::getline(file, line)) {
		std::istringstream ss(line);
		ss >> firstCol >> word;
		
		this->vitalsHeader.push_back(word);
	}
	this->vitalsHeader[0] = "timeStamp";

	file.close();
}

File: src/ros2_ws/src/logger/logger/logger_text.py
import os
import zipfile
from pathlib import Path
from typing import Tuple

from geometry_msgs.msg import PointStamped
from nav_msgs.msg import Odometry
from sensor_msgs.msg import Imu, NavSatFix, PointCloud2
from raspberrypi_vitals_msg.msg import sysinfo as SysInfo

from .logger_base import LoggerBase


class LoggerText(LoggerBase):
    """ROS2 port of the text logger: writes CSV-like text files per sensor, rotates, optional zip."""

    def __init__(self, output_folder: str, separator: str = ";") -> None:
        super().__init__("logger_text_node", output_folder, separator)
        self.tmp_dir = Path("/tmp")
        self.files = {}
        self.last_gnss_ts = 0
        self.last_imu_ts = 0
        self.last_sonar_ts = 0
        self.last_lidar_ts = 0
        self.last_speed_ts = 0
        self.last_vitals_ts = 0
        self.date_string = ""

    # LoggerBase hooks
    def init_logging(self):
        self.date_string = self._timestamp_prefix()
        self.files = {
            "gnss": self._open_file(self.date_string + "_gnss.txt", f"Timestamp{self.separator}Longitude{self.separator}Latitude{self.separator}EllipsoidalHeight{self.separator}Status{self.separator}Service\n"),
            "imu": self._open_file(self.date_string + "_imu.txt", f"Timestamp{self.separator}Heading{self.separator}Pitch{self.separator}Roll\n"),
            "sonar": self._open_file(self.date_string + "_sonar.txt", f"Timestamp{self.separator}Depth\n"),
            "lidar": self._open_file(self.date_string + "_lidar.txt", f"Timestamp{self.separator}Points\n"),
            "speed": self._open_file(self.date_string + "_speed.txt", f"Timestamp{self.separator}Speed\n"),
            "vitals": self._open_file(self.date_string + "_vitals.txt", f"Timestamp{self.separator}CPU Temp{self.separator}CPU Used{self.separator}FREE RAM{self.separator}FREE HDD{self.separator}Uptime{self.separator}Humidity{self.separator}System Temp{self.separator}Supply Voltage\n"),
        }

    def finalize_logging(self):
        target_dir = Path(self.output_folder)
        target_dir.mkdir(parents=True, exist_ok=True)
        for name, f in list(self.files.items()):
            try:
                fname = Path(f.name).name
                f.close()
                src = Path(f.name)
                dest = target_dir / fname
                try:
                    src.replace(dest)
                except Exception:
                    pass
            except Exception:
                pass
        self.files = {}
        # Zip logs
        if self.date_string:
            zip_name = target_dir / f"{self.date_string}.zip"
            with zipfile.ZipFile(zip_name, "w", compression=zipfile.ZIP_DEFLATED) as zf:
                for p in target_dir.iterdir():
                    if p.is_file() and p.name.startswith(self.date_string):
                        zf.write(p, arcname=p.name)

    def save_gnss(self, msg: NavSatFix):
        ts = msg.header.stamp.sec * 1_000_000 + msg.header.stamp.nanosec // 1_000
        if ts <= self.last_gnss_ts or "gnss" not in self.files:
            return
        f = self.files["gnss"]
        f.write(f"{msg.header.stamp.sec}.{msg.header.stamp.nanosec:09d}{self.separator}{msg.longitude:.10f}{self.separator}{msg.latitude:.10f}{self.separator}{msg.altitude:.3f}{self.separator}{msg.status.status}{self.separator}{msg.status.service}\n")
        f.flush()
        self.last_gnss_ts = ts

    def save_imu(self, msg: Imu):
        ts = msg.header.stamp.sec * 1_000_000 + msg.header.stamp.nanosec // 1_000
        if ts <= self.last_imu_ts or "imu" not in self.files:
            return
        f = self.files["imu"]
        f.write(f"{msg.header.stamp.sec}.{msg.header.stamp.nanosec:09d}{self.separator}{msg.orientation.x:.3f}{self.separator}{msg.orientation.y:.3f}{self.separator}{msg.orientation.z:.3f}\n")
        f.flush()
        self.last_imu_ts = ts

    def save_depth(self, msg: PointStamped):
        ts = msg.header.stamp.sec * 1_000_000 + msg.header.stamp.nanosec // 1_000
        if ts <= self.last_sonar_ts or "sonar" not in self.files:
            return
        f = self.files["sonar"]
        f.write(f"{msg.header.stamp.sec}.{msg.header.stamp.nanosec:09d}{self.separator}{msg.point.z:.3f}\n")
        f.flush()
        self.last_sonar_ts = ts

    def save_lidar(self, msg: PointCloud2):
        ts = msg.header.stamp.sec * 1_000_000 + msg.header.stamp.nanosec // 1_000
        if ts <= self.last_lidar_ts or "lidar" not in self.files:
            return
        f = self.files["lidar"]
        f.write(f"{msg.header.stamp.sec}.{msg.header.stamp.nanosec:09d}{self.separator}points\n")
        f.flush()
        self.last_lidar_ts = ts

    def save_speed(self, msg: Odometry):
        ts = msg.header.stamp.sec * 1_000_000 + msg.header.stamp.nanosec // 1_000
        if ts <= self.last_speed_ts or "speed" not in self.files:
            return
        f = self.files["speed"]
        f.write(f"{msg.header.stamp.sec}.{msg.header.stamp.nanosec:09d}{self.separator}{msg.twist.twist.linear.y:.3f}\n")
        f.flush()
        self.last_speed_ts = ts

    def save_vitals(self, msg: SysInfo):
        ts = msg.header.stamp.sec * 1_000_000 + msg.header.stamp.nanosec // 1_000
        if ts - self.last_vitals_ts < 60 * 1_000_000 or "vitals" not in self.files:
            return
        f = self.files["vitals"]
        f.write(
            f"{msg.header.stamp.sec}.{msg.header.stamp.nanosec:09d}{self.separator}"
            f"{msg.cputemp:.3f}{self.separator}{msg.cpuload:.3f}{self.separator}"
            f"{msg.freeram:.3f}{self.separator}{msg.freehdd:.3f}{self.separator}"
            f"{msg.uptime:.3f}{self.separator}{msg.humidity:.3f}{self.separator}"
            f"{msg.temperature:.3f}{self.separator}{msg.voltage:.3f}\n"
        )
        f.flush()
        self.last_vitals_ts = ts

    def transfer(self) -> Tuple[bool, str]:
        # Placeholder: remote transfer not implemented in this port
        return False, "transfer not implemented"

    # Helpers
    def _open_file(self, filename: str, header: str):
        path = self.tmp_dir / filename
        os.makedirs(self.tmp_dir, exist_ok=True)
        f = open(path, "a")
        f.write(header)
        return f

    def _timestamp_prefix(self) -> str:
        now = self.get_clock().now().to_msg()
        return f"{now.sec}"
